README.md
# Keyboard Status Color

Visual Studio Code Extension that changes the color of your keyboard while coding.
This extension only supports the `Tecware Phantom 87 TKL` keyboard.
> This extension isn't in the marketplace. You should clone this repository and run `npm install` and `npm run build`, after that, right-click the `.vsix` file in this directory and click `Install the VSIX extension`.

It changes the color of the keys `Esc`, `Right`, `Left`, `Up`, `Down`.

If there's an error globally, the color of the keys will be **red**.

If there's a warning globally, the color of the keys will be **yellow/orange**.

Otherwise, the color of the keys will be **blue**.

CHANGELOG.md
# Keyboard Status Color

### 1.0.0 - First release

src/util.ts
export const dataStart = [0x04, 0x01, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00];

export const dataEnd = [0x04, 0x02, 0x00, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00];

export const dataSettings = [0x04, 0x00, 0x00, 0x06, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00];

export const keys = {
    '0': [0xa8, 0x54, 0x00],
    '1': [0x8d, 0x39, 0x00],
    '2': [0x90, 0x3c, 0x00],
    '3': [0x93, 0x3f, 0x00],
    '4': [0x96, 0x42, 0x00],
    '5': [0x99, 0x45, 0x00],
    '6': [0x9c, 0x48, 0x00],
    '7': [0x9f, 0x4b, 0x00],
    '8': [0xa2, 0x4e, 0x00],
    '9': [0xa5, 0x51, 0x00],
    Esc: [0x57, 0x03, 0x00],
    F1: [0x5a, 0x06, 0x00],
    F2: [0x5d, 0x09, 0x00],
    F3: [0x60, 0x0c, 0x00],
    F4: [0x63, 0x0f, 0x00],
    F5: [0x66, 0x12, 0x00],
    F6: [0x69, 0x15, 0x00],
    F7: [0x6c, 0x18, 0x00],
    F8: [0x6f, 0x1b, 0x00],
    F9: [0x72, 0x1e, 0x00],
    F10: [0x75, 0x21, 0x00],
    F11: [0x78, 0x24, 0x00],
    F12: [0x7b, 0x27, 0x00],
    PrtSc: [0x93, 0x3e, 0x01],
    ScrLk: [0x96, 0x41, 0x01],
    Pause: [0x99, 0x44, 0x01],
    Tilde: [0x8a, 0x36, 0x00],
    Minus: [0xab, 0x57, 0x00],
    Equals: [0xae, 0x5a, 0x00],
    Backspace: [0x7b, 0x26, 0x01],
    Insert: [0x9f, 0x4a, 0x01],
    Home: [0xa2, 0x4d, 0x01],
    PgUp: [0xa5, 0x50, 0x01],
    Delete: [0xa8, 0x53, 0x01],
    End: [0xab, 0x56, 0x01],
    PgDn: [0xae, 0x59, 0x01],
    Tab: [0xbd, 0x69, 0x00],
    q: [0xc0, 0x6c, 0x00],
    w: [0xc3, 0x6f, 0x00],
    e: [0xc6, 0x72, 0x00],
    r: [0xc9, 0x75, 0x00],
    t: [0xcc, 0x78, 0x00],
    y: [0xcf, 0x7b, 0x00],
    u: [0xd2, 0x7e, 0x00],
    i: [0xd5, 0x81, 0x00],
    o: [0xd8, 0x84, 0x00],
    p: [0xdb, 0x87, 0x00],
    Bracket_l: [0xde, 0x8a, 0x00],
    Bracket_r: [0xe1, 0x8d, 0x00],
    Backslash: [0x14, 0xc0, 0x00],
    Up: [0x75, 0x20, 0x01],
    Left: [0x6f, 0x1a, 0x01],
    Down: [0x72, 0x1d, 0x01],
    Right: [0x78, 0x23, 0x01],
    Caps_Lock: [0xf0, 0x9c, 0x00],
    a: [0xf3, 0x9f, 0x00],
    s: [0xf6, 0xa2, 0x00],
    d: [0xf9, 0xa5, 0x00],
    f: [0xfc, 0xa8, 0x00],
    g: [0xff, 0xab, 0x00],
    h: [0x02, 0xae, 0x00],
    j: [0x05, 0xb1, 0x00],
    k: [0x08, 0xb4, 0x00],
    l: [0x0b, 0xb7, 0x00],
    Semicolon: [0x0e, 0xba, 0x00],
    Apostrophe: [0x11, 0xbd, 0x00],
    Return: [0x47, 0xf3, 0x00],
    Shift_l: [0x23, 0xcf, 0x00],
    z: [0x26, 0xd2, 0x00],
    x: [0x29, 0xd5, 0x00],
    c: [0x2c, 0xd8, 0x00],
    v: [0x2f, 0xdb, 0x00],
    b: [0x32, 0xde, 0x00],
    n: [0x35, 0xe1, 0x00],
    m: [0x38, 0xe4, 0x00],
    Comma: [0x3b, 0xe7, 0x00],
    Period: [0x3e, 0xea, 0x00],
    Slash: [0x41, 0xed, 0x00],
    Shift_r: [0x44, 0xf0, 0x00],
    Ctrl_l: [0x57, 0x02, 0x01],
    Super_l: [0x5a, 0x05, 0x01],
    Alt_l: [0x5d, 0x08, 0x01],
    Space: [0x60, 0x0b, 0x01],
    Alt_r: [0x63, 0x0e, 0x01],
    Fn: [0x66, 0x11, 0x01],
    Menu: [0x69, 0x14, 0x01],
    Ctrl_r: [0x6c, 0x17, 0x01],

    Int_Key: [0x4f, 0x3b, 0x01]
};

src/Keyboard.ts
import { dataSettings, dataStart, dataEnd, keys } from './util';
const usb = require('node-hid');

type KeyboardModes = 'custom' | 'fixed' | 'breathingCycle';

export default class Keyboard {
    usb = new usb.HID(0x0c45, 0x652f);

    setMode(mode: KeyboardModes) {
        const settings = [...dataSettings];
        switch (mode) {
            case 'custom':
                settings[1] = 0x1b;
                settings[8] = 0x14;
                break;
            case 'fixed':
                settings[1] = 0x0d;
                settings[8] = 0x06;
                break;
            case 'breathingCycle':
                settings[1] = 0x0b;
                settings[8] = 0x04;
                break;
        }

        this.usb.write(dataStart);
        this.usb.write(settings);
        this.usb.write(dataEnd);
    }

    setColor(red: number, green: number, blue: number) {
        const firstSettings = [...dataSettings];
        const secondSettings = [...dataSettings];
        firstSettings[1] = 0x0b;
        firstSettings[5] = 0x04;

        secondSettings[2] = 0x02;
        secondSettings[4] = 0x03;
        secondSettings[5] = 0x05;
        secondSettings[8] = red;
        secondSettings[9] = green;
        secondSettings[10] = blue;

        this.usb.write(dataStart);
        this.usb.write(firstSettings);
        this.usb.write(secondSettings);
        this.usb.write(dataEnd);
    }

    setKeysColor(keyss: (keyof typeof keys)[], red: number, green: number, blue: number) {
        this.usb.write(dataStart);

        keyss.forEach((key) => {
            const settings = [...dataSettings];

            settings[2] = 0x02;
            settings[3] = 0x11;
            settings[4] = 0x03;

            settings[1] = keys[key][0];
            settings[5] = keys[key][1];
            settings[6] = keys[key][2];

            settings[8] = red;
            settings[9] = green;
            settings[10] = blue;

            this.usb.write(settings);
        });

        this.usb.write(dataEnd);
    }

    setAllKeysColor(red: number, green: number, blue: number) {
        this.usb.write(dataStart);

        (Object.keys(keys) as (keyof typeof keys)[]).forEach((key) => {
            const settings = [...dataSettings];

            settings[2] = 0x02;
            settings[3] = 0x11;
            settings[4] = 0x03;

            settings[1] = keys[key][0];
            settings[5] = keys[key][1];
            settings[6] = keys[key][2];

            settings[8] = red;
            settings[9] = green;
            settings[10] = blue;

            this.usb.write(settings);
        });

        this.usb.write(dataEnd);
    }
}

src/extension.ts
import Keyboard from './Keyboard';
import * as vscode from 'vscode';
import { keys } from './util';

const keyboard = {
	_keyboard: new Keyboard(),
	_status: (null as vscode.DiagnosticSeverity | null),
	set status(value: vscode.DiagnosticSeverity | null) {
		if (this._status === value) return;
		this._status = value;
		this.onChangeStatus();
	},
	get status() {
		return this._status;
	},
	onChangeStatus() {
		/* const keyss: (keyof typeof keys)[] = [
			'Esc', 'Tab', 'Shift_l', 'Caps_Lock', 'Ctrl_l', 'Tilde', 'Super_l',
			'F12', 'Backslash', 'Shift_r', 'Ctrl_r', 'Menu', 'Return', 'Backspace',
			'Alt_l', 'Alt_r', 'Fn', 'Space'
		]; */
		/* const keyss = Object.keys(keys).filter((_, i) => i % 2 === 0) as (keyof typeof keys)[]; */
		const keyss: (keyof typeof keys)[] = ['Esc', 'Left', 'Down', 'Right', 'Up'];
		switch (this._status) {
			case vscode.DiagnosticSeverity.Error:
				this._keyboard.setKeysColor(keyss, 0xFF, 0x10, 0x10);
				break;
			case vscode.DiagnosticSeverity.Warning:
				this._keyboard.setKeysColor(keyss, 0xEA, 0x46, 0x00);
				break;
			default:
				this._keyboard.setKeysColor(keyss, 0x10, 0x10, 0xFF);
				break;
		}
	},
	init() {
		this._keyboard.setMode('custom');
		this._keyboard.setAllKeysColor(0x10, 0x10, 0xFF);
	}
};

export function activate() {
	const statusBar = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left);
	statusBar.text = '...';
	statusBar.show();
	keyboard.init();
	statusBar.tooltip = 'Keyboard working.';
	statusBar.text = '😋';

	vscode.languages.onDidChangeDiagnostics(() => {
		const diagnostics = vscode.languages.getDiagnostics()
			.map(([, diagnostic]) => diagnostic).flat();
		if (diagnostics.some((d) => d.severity === vscode.DiagnosticSeverity.Error))
			return (keyboard.status = vscode.DiagnosticSeverity.Error);
		if (diagnostics.some((d) => d.severity === vscode.DiagnosticSeverity.Warning))
			return (keyboard.status = vscode.DiagnosticSeverity.Warning);
		return (keyboard.status = null);
	});
}

export function deactivate() {
	keyboard._keyboard.setMode('breathingCycle');
}
